core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -0,0 +1,659 @@
+   *
+   * @param key the key to create
+   * @return true if the operation was successful
+   */
+  protected abstract boolean createEmptyObject(String key);
+
+  /**
+   * Creates an {@link OutputStream} for object uploads.
+   *
+   * @param key ufs key including scheme and bucket
+   * @throws IOException if failed to create stream
+   * @return new OutputStream
+   */
+  protected abstract OutputStream createObject(String key) throws IOException;
+
+  /**
+   * Appends the directory suffix to the key.
+   *
+   * @param key the key to convert
+   * @return key as a directory path
+   */
+  protected String convertToFolderName(String key) {
+    // Strips the slash if it is the end of the key string. This is because the slash at
+    // the end of the string is not part of the Object key.
+    key = CommonUtils.stripSuffixIfPresent(key, PATH_SEPARATOR);
+    return key + getFolderSuffix();
+  }
+
+  /**
+   * Copies an object to another key.
+   *
+   * @param src the source key to copy
+   * @param dst the destination key to copy to
+   * @return true if the operation was successful, false otherwise
+   */
+  protected abstract boolean copyObject(String src, String dst) throws IOException;
+
+  /**
+   * Internal function to delete a key.
+   *
+   * @param key the key to delete
+   * @return true if successful, false if an exception is thrown
+   */
+  protected abstract boolean deleteObject(String key) throws IOException;
+
+  /**
+   * Gets the metadata associated with a non-root path if it represents a folder. This method will
+   * return null if the path is not a folder. If the path exists as a prefix but does not have the
+   * folder dummy file, a folder dummy file will be created.
+   *
+   * @param path the path to get the metadata for
+   * @return the metadata of the folder, or null if the path does not represent a folder
+   */
+  protected ObjectStatus getFolderMetadata(String path) {
+    Preconditions.checkArgument(!isRoot(path));
+    String keyAsFolder = convertToFolderName(stripPrefixIfPresent(path));
+    ObjectStatus meta = getObjectStatus(keyAsFolder);
+    if (meta == null) {
+      String dir = stripPrefixIfPresent(path);
+      // Check if anything begins with <folder_path>/
+      try {
+        ObjectListingResult objs = getObjectListing(dir, true);
+        // If there are, this is a folder and we can create the necessary metadata
+        if (objs != null && objs.getObjectNames() != null && objs.getObjectNames().length > 0) {
+          mkdirsInternal(dir);
+          meta = getObjectStatus(keyAsFolder);
+        }
+      } catch (IOException e) {
+        meta = null;
+      }
+    }
+    return meta;
+  }
+
+  /**
+   * Get metadata information about object.
+   *
+   * @param key ufs key to get metadata for
+   * @return {@link ObjectStatus} if key exists and successful, otherwise null
+   */
+  protected abstract ObjectStatus getObjectStatus(String key);
```

Comment:
Seems like the method is only used for files. I don't know if we want to have a method for directories. If we do, I suggest we split it up into a different function instead of overloading this one.